reservation/manager.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return err
+	}
+
+	// Create the reservation state machine.
+	reservationFSM := NewFSM(
+		m.runCtx, m.cfg,
+	)
+
+	initContext := &InitReservationContext{
+		ReservationID: reservationID,
+		ServerPubkey:  serverKey,
+		Value:         btcutil.Amount(reservationRequest.Value),
+		Expiry:        reservationRequest.Expiry,
+		HeightHint:    uint32(m.currentHeight),
+	}
+
+	// Send the init event to the state machine.
+	err = reservationFSM.SendEvent(OnServerRequest, initContext)
+	if err != nil {
+		return err
+	}
+
+	// We'll now wait for the reservation to be in the state where it is
+	// waiting to be confirmed.
+	err = reservationFSM.DefaultObserver.WaitForState(
+		m.runCtx, time.Minute, WaitForConfirmation,
+		time.Second,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *Manager) RegisterReservationNotifications(
+	ctx context.Context) (chan *reservationrpc.
+	ServerReservationNotification, chan error, error) {
+
+	// In order to create a valid lsat we first are going to call
+	// the list reservations method.
+	_, err := r.cfg.ReservationClient.FetchL402(
+		ctx, &reservationrpc.FetchL402Request{},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// We'll now subscribe to the reservation notifications.
+	reservationStream, err := r.cfg.ReservationClient.
+		ReservationNotificationStream(
+			ctx, &reservationrpc.ReservationNotificationRequest{},
+		)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reservationChan := make(
+		chan *reservationrpc.ServerReservationNotification,
+	)
+	errChan := make(chan error, 1)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+
+			default:
+				reservationRes, err := reservationStream.Recv()
+				if err != nil {
+					errChan <- err
+					return
+				}
+				reservationChan <- reservationRes
+			}
+		}
+	}()
+
+	return reservationChan, errChan, nil
+}
+
+// RecoverReservations tries to recover all reservations that are still active
+// from the database.
+func (m *Manager) RecoverReservations(ctx context.Context) error {
+	reservations, err := m.cfg.Store.ListReservations(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, reservation := range reservations {
+		if isFinalState(reservation.State) {
+			continue
+		}
+
+		log.Debugf("Recovering reservation %x", reservation.ID)
+
+		fsmCtx := context.WithValue(ctx, reservation.ID, nil)
+
+		reservationFSM := NewFSMFromReservation(
+			fsmCtx, m.cfg, reservation,
+		)
+
+		m.activeReservations[reservation.ID] = reservationFSM
+
+		go func() {
+			err = reservationFSM.SendEvent(OnRecover, nil)
+			if err != nil {
+				log.Errorf("Error sending recover event: %v, "+
+					"state: %v", err,
+					reservationFSM.reservation.State)
+			}
+		}()
+	}
+
+	return nil
+}
+
+// GetReservations retrieves all reservations from the database.
+func (m *Manager) GetReservations(ctx context.Context) ([]*Reservation,
+	error) {
+
+	if m == nil {
```

Comment:
What's the benefit of checking these members (which should be set at this point) vs just letting the program crash instead?